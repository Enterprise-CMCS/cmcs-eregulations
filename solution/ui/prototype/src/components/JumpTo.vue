<template>
    <div class="jump-to">
        <div v-if="header !== ''" class="jump-to-label">
            {{ header }}
        </div>

        <form @submit.prevent="formSubmit">
            <input name="-version" type="hidden" required value="" />

            <div class="jump-to-input">
                <select
                    v-if="defaultTitle !== ''"
                    name="title"
                    class="ds-c-field"
                    required
                    v-model="selectedTitle"
                >
                    <option value="" disabled selected>Title</option>
                    <option value="42">42</option>
                </select>
                ยง
                <select
                    name="part"
                    class="ds-c-field"
                    aria-label="Regulation part number"
                    v-model="selectedPart"
                    required
                >
                    <template v-if="partNames">
                        <option value="" disable selected>Part</option>
                        <option
                            v-for="partName in partNames"
                            :value="partName"
                            :key="partName"
                        >
                            {{ partName }}
                        </option>
                    </template>
                </select>
                <span class="dot">.</span>
                <input
                    class="number-box ds-c-field"
                    name="section"
                    placeholder=""
                    type="text"
                    pattern="\d+"
                    title="Regulation section number, i.e. 111"
                    aria-label="Regulation section number, i.e. 111"
                />
            </div>

            <input class="submit" type="submit" value="Go" />
        </form>
    </div>
</template>

<script>
import { getPartNames } from "@/utilities/api";

export default {
    name: "JumpTo",
    props: {
        header: {
            type: String,
            required: false,
        },
        defaultTitle: {
            type: Number,
            required: false,
        },
    },

    data() {
        return {
            partNames: null,
            selectedPart: "",
            selectedTitle: "",
        };
    },

    async created() {
        try {
            this.partNames = await getPartNames();
        } catch (error) {
            console.error(error);
        }
    },

    methods: {
        formSubmit() {
            this.$router.push({
                name: "part",
                params: { title: this.selectedTitle, part: this.selectedPart },
            });
        },
    },
};
</script>

<style scoped>
* {
    box-sizing: border-box;
}

.jump-to .dot {
    margin: 0 5px;
}

.jump-to .jump-to-input select {
    
    border: #d6d7d9 1px solid;
}
.jump-to-input {
    padding: 10px;
}

.jump-to input.submit {
    border: solid 1px #d6d7d9;
    color: #a3a3a3;
    background-color: white;
}

.jump-to .number-box {
    border: #d6d7d9 1px solid;
}
</style>
