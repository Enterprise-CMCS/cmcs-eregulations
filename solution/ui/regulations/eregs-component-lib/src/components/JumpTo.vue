<template>
<div>
    <div class="jump-to-label">Jump to Regulation Section</div>
    <div class="form">
        <div class="jump-to-input">
            <select
                v-if="defaultTitle === ''"
                v-model="selectedTitle"
                name="title"
                class="ds-c-field"
                required
            >
                <option value="" disabled selected>Title</option>
                    <option
                        v-for="title in titles"
                        :key=title
                        :value=title
                    >
                        {{ title }} CFR
                    </option>
            </select>
                ยง
            <select
                v-model="selectedPart"
                name="part"
                class="ds-c-field"
                aria-label="Regulation part number"
                required
                :disabled="!selectedTitle"
            >
                <option value="" disable selected>Part</option>
                <option
                    v-for="part in filteredParts"
                    :key=part
                    :value=part
                >
                    {{ part }}
                </option>
            </select>
            <span class="dot">.</span>
            <input
                v-model="selectedSection"
                class="number-box ds-c-field"
                name="section"
                placeholder=""
                type="text"
                pattern="\d+"
                title="Regulation section number, i.e. 111"
                aria-label="Regulation section number, i.e. 111"
            />
        </div>
        <input class="submit" :class="{active: isActive}" type="submit" value="Go" @click="getLink" /></div>

</div>
</template>
<script>

import { getTitles, getParts } from "../api";

export default {
    name: "JumpTo",

    props: {
        apiurl: {
            type: String,
            required: true,
        },
        title: {
            type: String,
            required: false,
            default: ""
        },
        part: {
            type: String,
            required: false,
            default: ""
        },
    },
    data() {
        return {
            partNames: null,
            titles: [],
            selectedPart: "",
            defaultTitle:'',
            selectedTitle: "",
            selectedSection: "",
            hideTitle: false,
            filteredParts:[],
            isActive: false,
            parts:[],
            link: "",
        };
    },

    async created() {
        this.titles = await getTitles(this.apiurl);
        if(this.titles.length === 1){
            this.selectedTitle = this.titles[0];
            this.defaultTitle = this.selectedTitle;
            this.hideTitle= true;
        }
        if(this.title!==""){
            this.selectedTitle=this.title;
            this.hideTitle=true;
        }
        if(this.part !==""){
            this.selectedPart=this.part;
            this.isActive=true;
        }
    },

    watch:{
        selectedTitle(title) {
            if(title === ""){
                this.selectedPart=""
                this.isActive=false;
            }
            else{
                this.getParts(title)
            }
        },
    
        selectedPart(part) {
            if(part !== ""){
                this.isActive=true;
            }
            else{
                this.isActive=false;
            }
        }
    },

    methods: {
        async getParts(title) {
            const partsList = await getParts(this.apiurl, title)
            this.filteredParts = partsList.map((part) => part.name);
        },
        getLink(){
            if(this.isActive){
                const link = `/${this.selectedTitle}/${this.selectedPart}/${this.selectedSection}`
                window.location.href = link
            }
        }
    }    
};
</script>
