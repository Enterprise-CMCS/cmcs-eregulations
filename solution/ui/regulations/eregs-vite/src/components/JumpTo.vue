<template>
    <div>
        <form @submit.prevent="getLink">
            <div class="jump-to-input">
                <select
                    v-if="defaultTitle === ''"
                    id="jumpToTitle"
                    v-model="selectedTitle"
                    name="title"
                    aria-label="Regulation title number"
                    class="ds-c-field"
                    required
                >
                    <option
                        value=""
                        disabled
                        selected
                    >
                        Title
                    </option>
                    <option
                        v-for="listedTitle in titles"
                        :key="listedTitle"
                        :value="listedTitle"
                    >
                        {{ listedTitle }} CFR
                    </option>
                </select>
                <span class="section-symbol">ยง</span>
                <select
                    id="jumpToPart"
                    v-model="selectedPart"
                    name="part"
                    class="ds-c-field"
                    aria-label="Regulation part number"
                    required
                    :disabled="!selectedTitle"
                >
                    <option
                        value=""
                        disable
                        selected
                    >
                        Part
                    </option>
                    <option
                        v-for="listedPart in filteredParts"
                        :key="listedPart"
                        :value="listedPart"
                    >
                        {{ listedPart }}
                    </option>
                </select>
                <span class="dot">.</span>
                <input
                    id="jumpToSection"
                    v-model="selectedSection"
                    class="number-box ds-c-field"
                    name="section"
                    placeholder=""
                    type="text"
                    pattern="\d+"
                    title="Regulation section number, i.e. 111"
                    aria-label="Regulation section number, i.e. 111"
                >
                <input
                    id="jumpBtn"
                    class="submit"
                    :class="{ active: isActive }"
                    type="submit"
                    value="Go"
                >
            </div>
        </form>
    </div>
</template>
<script>
import { getTitles, getParts as fetchParts } from "utilities/api.js";

export default {
    name: "JumpTo",

    props: {
        apiUrl: {
            type: String,
            required: true,
        },
        title: {
            type: String,
            required: false,
            default: "",
        },
        part: {
            type: String,
            required: false,
            default: "",
        },
        homeUrl: {
            type: String,
            default: "/",
        },
    },
    data() {
        return {
            titles: [],
            selectedPart: "",
            defaultTitle: "",
            selectedTitle: "",
            selectedSection: "",
            hideTitle: false,
            filteredParts: [],
            isActive: false,
            parts: [],
            link: "",
        };
    },

    async created() {
        // When title 45 or another title is added uncomment line below, and remove the hardcoded
        this.titles = await getTitles({ apiUrl: this.apiUrl });
        if (this.titles.length === 1) {
            this.selectedTitle = this.titles[0];
            this.defaultTitle = this.selectedTitle;
            this.hideTitle = true;
        }
        if (this.title !== "") {
            this.selectedTitle = this.title;
            this.hideTitle = true;
        }
        if (this.part !== "") {
            this.selectedPart = this.part;
            this.isActive = true;
        }
    },

    watch: {
        selectedTitle(title) {
            if (title === "") {
                this.selectedPart = "";
                this.isActive = false;
            } else {
                this.getParts(title);
            }
        },

        selectedPart(part) {
            if (part !== "") {
                this.isActive = true;
            } else {
                this.isActive = false;
            }
        },
    },

    methods: {
        async getParts(title) {
            const partsList = await fetchParts({ title, apiUrl: this.apiUrl });
            // temporarily filter part 75. See EREGCSC-1397
            this.filteredParts = partsList
                .map((part) => part.name)
                .filter((part) => part !== "75");
        },
        getLink() {
            let link = `${this.homeUrl}goto/?title=${this.selectedTitle}&part=${this.selectedPart}`;
            if (this.selectedSection !== "") {
                link += `&section=${this.selectedSection}&-version='latest'`;
            }

            window.location.href = link;
        },
    },
};
</script>
