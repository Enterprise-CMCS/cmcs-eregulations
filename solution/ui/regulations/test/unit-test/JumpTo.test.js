import { render, screen, fireEvent } from "@testing-library/vue";
import JumpTo from "vite/components/JumpTo.vue";
import { describe, it, expect } from "vitest";

import flushPromises from "flush-promises";

describe("Jump to", () => {
    it("Gets a list of parts after title is selected", async () => {
        const wrapper = render(JumpTo, {
            props: {
                apiUrl: "http://localhost:8000/"
            }
        });
        await flushPromises();
        const titles = ["", 42, 45];

        const titleDropdown = await wrapper.findByLabelText(
            "Regulation title number"
        );
        let partDropdown = await screen.findByLabelText(
            "Regulation part number"
        );
        expect(partDropdown._vOptions).toStrictEqual([""]);
        expect(titleDropdown._vOptions).toStrictEqual(titles);
        fireEvent.change(titleDropdown, { target: { value: 42 } });

        await flushPromises();
        let parts = ["", "400", "430", "431", "432", "433"];
        partDropdown = await screen.findByLabelText("Regulation part number");
        expect(partDropdown._vOptions).toStrictEqual(parts);
        parts = ["", "95", "155"];
        fireEvent.change(titleDropdown, { target: { value: 45 } });
        await flushPromises();
        expect(partDropdown._vOptions).toStrictEqual(parts);
        fireEvent.change(partDropdown, { target: { value: "" } });
    });
    it("Checks to see if the snap shot matches", async () => {
        const wrapper = render(JumpTo, {
            props: {
                apiUrl: "http://localhost:8000/"
            }
        });
        expect(wrapper).toMatchSnapshot();
    });
});
