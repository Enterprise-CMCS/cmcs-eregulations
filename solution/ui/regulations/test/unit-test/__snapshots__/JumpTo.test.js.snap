// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Jump to > Checks to see if the snap shot matches 1`] = `
{
  "baseElement": <body>
    <div>
      <div>
        <form>
          <div
            class="jump-to-input"
          >
            <select
              aria-label="Regulation title number"
              class="ds-c-field"
              id="jumpToTitle"
              name="title"
              required="required"
            >
              <option
                disabled="disabled"
                selected="selected"
                value=""
              >
                Title
              </option>
            </select>
            <span
              class="section-symbol"
            >
              ยง
            </span>
            <select
              aria-label="Regulation part number"
              class="ds-c-field"
              disabled="disabled"
              id="jumpToPart"
              name="part"
              required="required"
            >
              <option
                disable=""
                selected="selected"
                value=""
              >
                Part
              </option>
            </select>
            <span
              class="dot"
            >
              .
            </span>
            <input
              aria-label="Regulation section number, i.e. 111"
              class="number-box ds-c-field"
              id="jumpToSection"
              name="section"
              pattern="\\\\d+"
              placeholder=""
              title="Regulation section number, i.e. 111"
              type="text"
            />
            <input
              class="submit"
              id="jumpBtn"
              type="submit"
              value="Go"
            />
          </div>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <form>
        <div
          class="jump-to-input"
        >
          <select
            aria-label="Regulation title number"
            class="ds-c-field"
            id="jumpToTitle"
            name="title"
            required="required"
          >
            <option
              disabled="disabled"
              selected="selected"
              value=""
            >
              Title
            </option>
          </select>
          <span
            class="section-symbol"
          >
            ยง
          </span>
          <select
            aria-label="Regulation part number"
            class="ds-c-field"
            disabled="disabled"
            id="jumpToPart"
            name="part"
            required="required"
          >
            <option
              disable=""
              selected="selected"
              value=""
            >
              Part
            </option>
          </select>
          <span
            class="dot"
          >
            .
          </span>
          <input
            aria-label="Regulation section number, i.e. 111"
            class="number-box ds-c-field"
            id="jumpToSection"
            name="section"
            pattern="\\\\d+"
            placeholder=""
            title="Regulation section number, i.e. 111"
            type="text"
          />
          <input
            class="submit"
            id="jumpBtn"
            type="submit"
            value="Go"
          />
        </div>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "emitted": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "html": [Function],
  "isUnmounted": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "unmount": [Function],
  "updateProps": [Function],
}
`;
